src/vset.h

```diff
@@ -0,0 +1,317 @@
+ *        We search the first 2 following entries which belong to different lwo granularity time windows,
+ *        but as close as possible to the middle of the vector:
+ *            [entry1, entry7, ..., entry13]
+ *                          ↑
+ *                         split (first where get_bucket_ts(entry) > min_ts)
+ *
+ *     2. Create two vectors:
+ *            bucket A -> [entry1..entry6]  with key = "000032"
+ *            bucket B -> [entry7..entry13] with key = "000048"
+ *
+ *     3. Insert both back to the RAX.
+ *
+ *-----------------------------------------------------------------------------
+ * Bucket Lifecycle
+ *-----------------------------------------------------------------------------
+ *
+ *     NONE
+ *       |
+ *       v
+ *     SINGLE (1 entry)
+ *       |
+ *       v
+ *     VECTOR (sorted, up to 127)
+ *       |
+ *       v
+ *     RAX (holds multiple buckets, keyed by each bucket's end timestamp)
+ *     Bucket types within a RAX:
+ *
+ *                    SINGLE
+ *                      |
+ *                      v
+ *                    VECTOR (sorted, up to 127, can split
+ *                      |     into multiple vectors)
+ *                      |
+ *                      v
+ *                   HASHTABLE (only when a vector can't split)
+ *
+ *-----------------------------------------------------------------------------
+ * Entry Type Contract
+ *-----------------------------------------------------------------------------
+ *
+ * Users must supply a `volatileEntryType` implementation:
+ *
+ *     typedef struct {
+ *         sds (*entryGetKey)(const void *entry);      // get key
+ *         long long (*getExpiry)(const void *entry);  // get expiry
+ *         int (*expire)(void *db, void *o, void *entry); // trigger expiry
+ *     } volatileEntryType;
+ *
+ *-----------------------------------------------------------------------------
+ * Public API
+ *-----------------------------------------------------------------------------
+ *
+ * Create/Free:
+ *     void vsetInit(vset *set);
+ *     void vsetClear(vset *set);
+ *
+ * Mutation:
+ *     bool vsetAddEntry(vset *set, vsetGetExpiryFunc getExpiry, void *entry);
+ *     bool vsetRemoveEntry(vset *set, vsetGetExpiryFunc getExpiry, void *entry);
+ *     bool vsetUpdateEntry(vset *set, vsetGetExpiryFunc getExpiry, void *old_entry,
+ *                                void *new_entry, long long old_expiry,
+ *                                long long new_expiry);
+ *
+ * Expiry Retrieval/Removal:
+ *     long long vsetEstimatedEarliestExpiry(vset *set, vsetGetExpiryFunc getExpiry);
+ *     size_t vsetPopExpired(vset *set, vsetGetExpiryFunc getExpiry, vsetExpiryFunc expiryFunc, mstime_t now, size_t max_count, void *ctx);
+ *
+ * Utilities:
+ *     bool vsetIsEmpty(vset *set);
+ *
+ * Iteration:
+ *     void vsetStart(vset *set, vsetIterator *it);
+ *     bool vsetNext(vsetIterator *it, void **entryptr);
+ *     void vsetStop(vsetIterator *it);
```

Comment:
This documentation adds no value.  It provides nothing more than the declarations below.